File: projects/9ee52fd7-13e2-4fa0-9830-79d5543e1e32/archive/20230815_134247/workspace/DataMapper.js
class DataMapper {
  map(source, mapping) {
    const target = {};

    for (const [key, value] of Object.entries(mapping)) {
      const sourceValue = this.getValue(source, value.input.sourcePath);
      this.setValue(target, value.output.path, sourceValue);
    }

    return target;
  }

  getValue(object, path) {
    return path.split('.').reduce((obj, key) => obj[key], object);
  }

  setValue(object, path, value) {
    const keys = path.split('.');
    keys.slice(0, -1).reduce((obj, key) => obj[key] = obj[key] || {}, object)[keys.pop()] = value;
  }
}

module.exports = DataMapper;

File: projects/9ee52fd7-13e2-4fa0-9830-79d5543e1e32/archive/20230815_144528/workspace/data_mapper.py
class DataMapper:
    def map_data(self, source_data, mapping_json):
        mapped_data = {}
        for key, mapping in mapping_json.items():
            # Extract the source value
            source_path = mapping['input']['sourcePath'].split('.')
            source_value = source_data
            for path in source_path:
                source_value = source_value[path]

            # Apply formulas to the source value
            for formula in mapping['input'].get('formulas', []):
                source_value = self.apply_formula(source_value, formula)

            # Set the target value
            target_path = mapping['output']['outputPath'].split('.')
            target = mapped_data
            for path in target_path[:-1]:
                if path not in target:
                    target[path] = {}
                target = target[path]
            target[target_path[-1]] = source_value

        return mapped_data

    def apply_formula(self, value, formula):
        # Placeholder for applying a formula to a value
        # In a real application, this method should apply the formula to the value and return the result
        return value
